### PR TITLE
[Enhancement](inverted index) support inverted index format v2

### PR DESCRIPTION
**Pull request by airborne12**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/segment_writer.cpp [hunk 1]
```diff
@@ -206,15 +214,17 @@ Status SegmentWriter::init(const std::vector<uint32_t>& col_ids, bool has_key) {
             skip_inverted_index = true;
         }
         // indexes for this column
-        opts.indexes = _tablet_schema->get_indexes_for_column(column);
+        opts.indexes = std::move(_tablet_schema->get_indexes_for_column_by_copy(column));
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/segment_writer.cpp:217]: why copy needed here?

File: be/src/olap/rowset/beta_rowset.cpp [hunk 2]
```diff
@@ -92,20 +92,17 @@ Status BetaRowset::do_load(bool /*use_cache*/) {
...
-Status BetaRowset::get_inverted_index_size_by_index_id(int64_t index_id, size_t* index_size) {
+Status BetaRowset::get_inverted_index_size(size_t* index_size) {
     auto fs = _rowset_meta->fs();
     if (!fs || _schema == nullptr) {
         return Status::Error<INIT_FAILED>("get fs failed");
     }
     for (int seg_id = 0; seg_id < num_segments(); ++seg_id) {
         auto seg_path = segment_file_path(seg_id);
         int64_t file_size = 0;
-        const auto* index = _schema->get_inverted_index_with_index_id(index_id, "");
```
Comment (xiaokang) [inline be/src/olap/rowset/beta_rowset.cpp:103]: Is V1 index file size not calculated?

**Comment by airborne12** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/segment.cpp [hunk 3]
```diff
@@ -101,6 +102,9 @@ Segment::~Segment() {
 }
 
 Status Segment::_open() {
+    if (_tablet_schema->has_inverted_index()) {
+        RETURN_IF_ERROR(_open_inverted_index());
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/segment.cpp:106]: Lazy open may be necessary.

**Review by xiaokang — COMMENTED** [review]
table, rowset, segment related code reviewed

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

File: gensrc/proto/olap_file.proto [hunk 4]
```diff
@@ -345,6 +350,7 @@ message TabletSchemaPB {
     optional bool enable_single_replica_compaction = 22 [default=false];
     optional bool skip_write_index_on_load = 23 [default=false];
     repeated int32 cluster_key_idxes = 24;
+    optional InvertedIndexStorageFormatPB inverted_index_storage_format = 25 [default=V2];
```
[the remark this answers is not in the document]
Comment (airborne12) [reply gensrc/proto/olap_file.proto:353]: both needed

File: be/src/olap/rowset/segment_v2/inverted_index_compaction.h [hunk 5]
```diff
@@ -16,19 +16,23 @@
...
+class TabletIndex;
 namespace segment_v2 {
-Status compact_column(int32_t index_id, int src_segment_num, int dest_segment_num,
-                      std::vector<std::string> src_index_files,
-                      std::vector<std::string> dest_index_files, const io::FileSystemSPtr& fs,
-                      std::string index_writer_path, std::string tablet_path,
+class InvertedIndexFileWriter;
+class InvertedIndexFileReader;
+
+Status compact_column(int64_t index_id, std::vector<lucene::store::Directory*>& src_index_dirs,
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index_compaction.h:33]: Is int32_t not enough?

File: be/src/olap/rowset/segment_v2/inverted_index_compaction.cpp [hunk 6]
```diff
@@ -80,19 +59,19 @@ Status compact_column(int32_t index_id, int src_segment_num, int dest_segment_nu
     for (auto* d : src_index_dirs) {
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index_compaction.cpp:59]: If src_index_dirs and dst_index_dirs are not owned here, close() can also be moved to owner.

File: be/src/olap/task/index_builder.h [hunk 7]
```diff
@@ -59,6 +60,7 @@ class IndexBuilder {
 private:
     Status _write_inverted_index_data(TabletSchemaSPtr tablet_schema, int32_t segment_idx,
                                       vectorized::Block* block);
+    Status handle_inverted_index_update();
```
Comment (xiaokang) [inline be/src/olap/task/index_builder.h:63]: private function name: _handle_inverted_index_update()

Comment (xiaokang) [reply be/src/olap/task/index_builder.h:63]: it's not used any where.

File: be/src/olap/task/index_builder.cpp [hunk 8]
```diff
@@ -69,12 +69,28 @@ Status IndexBuilder::update_inverted_index_info() {
...
-                        t_inverted_index.index_id, &index_size));
-                total_index_size += index_size;
+                DCHECK_EQ(t_inverted_index.columns.size(), 1);
+                auto column_name = t_inverted_index.columns[0];
+                auto column_idx = output_rs_tablet_schema->field_index(column_name);
+                if (column_idx < 0) {
+                    LOG(WARNING) << "referenced column was missing. "
+                                 << "[column=" << column_name << " referenced_column=" << column_idx
+                                 << "]";
+                    continue;
```
Comment (xiaokang) [inline be/src/olap/task/index_builder.cpp:84]: mark for error handle

File: be/src/olap/task/index_builder.cpp [hunk 9]
```diff
@@ -69,12 +69,28 @@ Status IndexBuilder::update_inverted_index_info() {
...
+                }
+                auto column = output_rs_tablet_schema->column(column_idx);
+                const auto* index_meta = output_rs_tablet_schema->get_inverted_index(column);
+                if (index_meta == nullptr) {
+                    LOG(ERROR) << "failed to find column: " << column_name
+                               << " index_id: " << t_inverted_index.index_id;
+                    continue;
```
Comment (xiaokang) [inline be/src/olap/task/index_builder.cpp:91]: mark for error handle

**Review by xiaokang — COMMENTED** [review]
write and query code reviewed

**Comment by airborne12** [comment]
run buildall

File: be/src/olap/compaction.cpp [hunk 10]
```diff
@@ -582,53 +640,54 @@ void CompactionMixin::construct_skip_inverted_index(RowsetWriterContext& ctx) {
...
+                                 << " error:" << st;
+                    return false;
+                }
+
                 bool exists = false;
-                if (!fs->exists(inverted_index_src_file_path, &exists).ok()) {
-                    LOG(ERROR) << inverted_index_src_file_path << " fs->exists error";
+                if (!inverted_index_file_reader->index_file_exist(index_meta, &exists).ok()) {
+                    LOG(ERROR) << inverted_index_file_reader->get_index_file_path(index_meta)
+                               << " fs->exists error";
```
Comment (xiaokang) [inline be/src/olap/compaction.cpp:660]: index_file_exist error

File: be/src/olap/compaction.cpp [hunk 11]
```diff
@@ -582,53 +640,54 @@ void CompactionMixin::construct_skip_inverted_index(RowsetWriterContext& ctx) {
...
                 if (!exists) {
                     LOG(WARNING) << "tablet[" << _tablet->tablet_id() << "] column_unique_id["
-                                 << col_unique_id << "]," << inverted_index_src_file_path
+                                 << col_unique_id << "],"
+                                 << inverted_index_file_reader->get_index_file_path(index_meta)
                                  << " is not exists, will skip index compaction";
                     return false;
                 }
 
-                // check idx file size
```
Comment (qidaye) [inline be/src/olap/compaction.cpp:602]: Why remove this check?

File: be/src/olap/compaction.cpp [hunk 12]
```diff
@@ -490,29 +492,70 @@ Status CompactionMixin::do_inverted_index_compaction() {
...
     for (int i = 0; i < dest_segment_num; ++i) {
-        auto prefix = dest_rowset_id.to_string() + "_" + std::to_string(i);
-        dest_index_files[i] = prefix;
+        auto prefix = dest_rowset_id.to_string() + "_" + std::to_string(i) + ".dat";
+        auto inverted_index_file_reader = std::make_unique<InvertedIndexFileReader>(
+                fs, tablet_path, prefix, _cur_tablet_schema->get_inverted_index_storage_format());
+        bool open_idx_file_cache = false;
+        auto st = inverted_index_file_reader->init(config::inverted_index_read_buffer_size,
+                                                   open_idx_file_cache);
+        if (st.ok()) {
```
Comment (qidaye) [inline be/src/olap/compaction.cpp:532]: Add some comments to explain the logic.

File: be/src/olap/tablet_meta.cpp [hunk 13]
```diff
@@ -168,6 +170,18 @@ TabletMeta::TabletMeta(int64_t table_id, int64_t partition_id, int64_t tablet_id
...
+    switch (inverted_index_storage_format) {
+    case TInvertedIndexStorageFormat::V1:
+        schema->set_inverted_index_storage_format(InvertedIndexStorageFormatPB::V1);
+        break;
+    case TInvertedIndexStorageFormat::V2:
+        schema->set_inverted_index_storage_format(InvertedIndexStorageFormatPB::V2);
+        break;
+    default:
+        schema->set_inverted_index_storage_format(InvertedIndexStorageFormatPB::V2);
+        break;
```
Comment (qidaye) [inline be/src/olap/tablet_meta.cpp:182]: Maybe merge to one break?

File: gensrc/thrift/AgentService.thrift [hunk 14]
```diff
@@ -151,7 +159,7 @@ struct TCreateTabletReq {
     24: optional i64 time_series_compaction_file_count_threshold = 2000
     25: optional i64 time_series_compaction_time_threshold_seconds = 3600
     26: optional i64 time_series_compaction_empty_rowsets_threshold = 5
-
+    27: optional TInvertedIndexStorageFormat inverted_index_storage_format
```
Comment (qidaye) [inline gensrc/thrift/AgentService.thrift:162]: Default value?

File: be/src/olap/rowset/segment_v2/inverted_index_file_reader.cpp [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+    }
+    return Status::OK();
+}
+
+Result<InvertedIndexDirectoryMap> InvertedIndexFileReader::get_all_directories() {
+    InvertedIndexDirectoryMap res;
+    for (auto& index : _indices_entries) {
+        auto index_id = index.first.first;
+        auto index_suffix = index.first.second;
+        LOG(INFO) << "index_id:" << index_id << " index_suffix:" << index_suffix;
```
Comment (qidaye) [inline be/src/olap/rowset/segment_v2/inverted_index_file_reader.cpp:130]: debug level may be better.

**Review by xiaokang — COMMENTED** [review]
compaction reviewed

**Comment by airborne12** [comment]
run buildall

File: be/src/olap/task/index_builder.cpp [hunk 16]
```diff
@@ -67,39 +67,34 @@ Status IndexBuilder::update_inverted_index_info() {
...
-                } else {
-                    LOG(WARNING) << "rowset " << input_rowset_meta->rowset_id() << " " << size_type
-                                 << " size:" << disk_size
-                                 << " is less than index size:" << total_index_size;
-                }
-            };
-
-            size_t before_size = input_rowset_meta->total_disk_size();
-            update_disk_size(before_size, "total disk");
-            input_rowset_meta->set_total_disk_size(before_size);
```
Comment (xiaokang) [inline be/src/olap/task/index_builder.cpp:94]: mark: why delete

File: be/src/olap/task/index_builder.cpp [hunk 17]
```diff
@@ -137,6 +132,23 @@ Status IndexBuilder::update_inverted_index_info() {
...
+            std::string segment_filename = segment_full_path.filename().native();
+            auto segment_dir = segment_full_path.parent_path();
+            auto idx_file_reader = std::make_unique<InvertedIndexFileReader>(
+                    context.fs, segment_dir, segment_filename,
+                    output_rs_tablet_schema->get_inverted_index_storage_format());
+            auto st = idx_file_reader->init();
+            if (!st.ok() && !st.is<ErrorCode::INVERTED_INDEX_FILE_NOT_FOUND>()) {
+                return st;
+            }
+            _inverted_index_file_readers.emplace(
```
Comment (xiaokang) [inline be/src/olap/task/index_builder.cpp:147]: mark: why need index reader?

File: be/src/olap/task/index_builder.cpp [hunk 18]
```diff
@@ -175,7 +187,52 @@ Status IndexBuilder::update_inverted_index_info() {
 Status IndexBuilder::handle_single_rowset(RowsetMetaSharedPtr output_rowset_meta,
                                           std::vector<segment_v2::SegmentSharedPtr>& segments) {
     if (_is_drop_op) {
-        // delete invertd index file by gc thread when gc input rowset
+        std::string segment_dir = _tablet->tablet_path();
```
Comment (xiaokang) [inline be/src/olap/task/index_builder.cpp:190]: why not the same as before?

File: be/src/olap/task/index_builder.cpp [hunk 19]
```diff
@@ -189,6 +246,18 @@ Status IndexBuilder::handle_single_rowset(RowsetMetaSharedPtr output_rowset_meta
...
+            if (idx_file_reader_iter == _inverted_index_file_readers.end()) {
+                LOG(ERROR) << "idx_file_reader_iter" << output_rowset_meta->rowset_id() << ":"
+                           << seg_ptr->id() << " cannot be found";
+                continue;
+            }
+            auto dirs = DORIS_TRY(idx_file_reader_iter->second->get_all_directories());
+            auto inverted_index_file_writer = std::make_unique<InvertedIndexFileWriter>(
+                    fs, segment_dir, segment_filename,
+                    output_rowset_schema->get_inverted_index_storage_format());
+            RETURN_IF_ERROR(inverted_index_file_writer->initialize(dirs));
```
Comment (xiaokang) [inline be/src/olap/task/index_builder.cpp:260]: why initialize writer with reader dirs?

Comment (xiaokang) [reply be/src/olap/task/index_builder.cpp:260]: It's straightforward to init writers by index_meta as in normal writh path.

**Review by xiaokang — COMMENTED** [review]
build index reviewed

File: be/src/olap/rowset/segment_v2/inverted_index_writer.cpp [hunk 20]
```diff
@@ -616,21 +602,18 @@ class InvertedIndexColumnWriterImpl : public InvertedIndexColumnWriter {
     std::unique_ptr<lucene::analysis::Analyzer> _analyzer = nullptr;
     std::unique_ptr<lucene::util::Reader> _char_string_reader = nullptr;
     std::shared_ptr<lucene::util::bkd::bkd_writer> _bkd_writer = nullptr;
-    std::unique_ptr<DorisCompoundDirectory> _dir = nullptr;
-    std::string _segment_file_name;
-    std::string _directory;
-    io::FileSystemSPtr _fs;
+    DorisCompoundDirectory* _dir = nullptr;
```
[the remark this answers is not in the document]
Comment (airborne12) [reply be/src/olap/rowset/segment_v2/inverted_index_writer.cpp:603]: yes, in InvertedIndexDirectoryMap _indices_dirs

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Review by zzzxl1993 — APPROVED** [review]
LGTM

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Review by zzzxl1993 — APPROVED** [review]
LGTM

**Review by qidaye — APPROVED** [review]

**Pull request merged**